File: src/arsenals.py
"""Shared arsenal loader from data/Arsenals.csv.

Provides player metadata and pitch arsenal lookups, cached after first load.
"""

import csv
import os

_DEFAULT_PATH = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
                             "data", "Arsenals.csv")

_cache: dict | None = None


def _load(csv_path: str | None = None) -> dict:
    """Load and cache arsenals CSV. Returns dict keyed by player_id."""
    global _cache
    if _cache is not None:
        return _cache

    path = csv_path or _DEFAULT_PATH
    _cache = {}
    if not os.path.exists(path):
        return _cache

    try:
        with open(path, newline="") as f:
            reader = csv.DictReader(f)
            # Validate required columns
            required = {"player_id", "abbreviation", "pitch_type"}
            if not required.issubset(set(reader.fieldnames or [])):
                print(f"WARNING: {path} missing columns {required - set(reader.fieldnames or [])}")
                return _cache

            for row in reader:
                pid = row.get("player_id", "").strip()
                if not pid:
                    continue
                if pid not in _cache:
                    _cache[pid] = {
                        "player_name": row.get("player_name", "").strip() or None,
                        "pitcher_hand": row.get("pitcher_hand", "").strip() or None,
                        "arsenal": [],
                    }
                abbrev = row.get("abbreviation", "").strip()
                ptype = row.get("pitch_type", "").strip()
                if abbrev:
                    _cache[pid]["arsenal"].append({
                        "abbreviation": abbrev,
                        "pitch_type": ptype,
                    })
    except Exception as e:
        print(f"WARNING: Could not load arsenals from {path}: {e}")

    return _cache


def reload(csv_path: str | None = None) -> None:
    """Force reload of arsenals cache."""
    global _cache
    _cache = None
    _load(csv_path)


def get_player_name(player_id: str, csv_path: str | None = None) -> str | None:
    data = _load(csv_path)
    entry = data.get(player_id)
    return entry["player_name"] if entry else None


def get_player_hand(player_id: str, csv_path: str | None = None) -> str | None:
    data = _load(csv_path)
    entry = data.get(player_id)
    return entry["pitcher_hand"] if entry else None


def get_player_arsenal(player_id: str, csv_path: str | None = None) -> list[dict]:
    """Return list of {abbreviation, pitch_type} dicts for the player."""
    data = _load(csv_path)
    entry = data.get(player_id)
    return entry["arsenal"] if entry else []

File: web/lib/arsenals.ts
/**
 * Arsenal loader — fetches /data/Arsenals.csv and provides player lookups.
 * Cached after first fetch.
 */

interface ArsenalEntry {
  abbrev: string;
  pitchName: string;
}

interface PlayerMeta {
  playerName?: string;
  pitcherHand?: string;
}

interface PlayerRecord {
  playerName?: string;
  pitcherHand?: string;
  arsenal: ArsenalEntry[];
}

let loadPromise: Promise<Map<string, PlayerRecord>> | null = null;

function load(): Promise<Map<string, PlayerRecord>> {
  if (loadPromise) return loadPromise;

  loadPromise = (async () => {
    const map = new Map<string, PlayerRecord>();
    try {
      const res = await fetch("/data/Arsenals.csv");
      if (!res.ok) return map;
      const text = await res.text();
      const lines = text.trim().split("\n");
      if (lines.length < 2) return map;

      const header = lines[0].split(",").map((h) => h.trim());
      const idx = {
        player_id: header.indexOf("player_id"),
        player_name: header.indexOf("player_name"),
        pitcher_hand: header.indexOf("pitcher_hand"),
        pitch_type: header.indexOf("pitch_type"),
        abbreviation: header.indexOf("abbreviation"),
      };

      for (let i = 1; i < lines.length; i++) {
        const cols = lines[i].split(",").map((c) => c.trim());
        const pid = cols[idx.player_id] ?? "";
        if (!pid) continue;

        if (!map.has(pid)) {
          map.set(pid, {
            playerName: cols[idx.player_name] || undefined,
            pitcherHand: cols[idx.pitcher_hand] || undefined,
            arsenal: [],
          });
        }
        const abbrev = cols[idx.abbreviation] ?? "";
        if (abbrev) {
          map.get(pid)!.arsenal.push({
            abbrev,
            pitchName: cols[idx.pitch_type] ?? "",
          });
        }
      }
    } catch {
      // Arsenals.csv missing or unparseable — degrade gracefully
    }
    return map;
  })();

  return loadPromise;
}

export async function getPlayerMeta(
  playerId: string,
): Promise<PlayerMeta> {
  const data = await load();
  const rec = data.get(playerId);
  return {
    playerName: rec?.playerName,
    pitcherHand: rec?.pitcherHand,
  };
}

export async function getPlayerArsenal(
  playerId: string,
): Promise<ArsenalEntry[]> {
  const data = await load();
  return data.get(playerId)?.arsenal ?? [];
}
